### DISCUSSION TITLE
[A-L] (30 maggio 2024) Esercizio 1

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/progettazione-di-algoritmi#104] [Esercitazioni A-L]
![immagine](https://github.com/user-attachments/assets/5ef6cadb-0c6d-4b25-acd2-79a86c852f59)

**Comment by Elia-Belli** [comment]
Ripercorrendo il vettore dei padri a partire da un qualsiasi nodo possiamo calcolare se tra i suoi antenati c'è $u$ , per evitare di ricalcolare l'appartenenza al sottoalbero $T_u$ di nodi già analizzati, manteniamo un array $Vis[v]$ che per ogni nodo $v$ indica se appartiene a $T_u$ $(1)$, se non vi appartiene $(0)$ o se non è ancora stato stabilito $(-1)$ : un figlio che non conosce ancora la sua discendenza effettuerà una chiamata ricorsiva al padre ed al ritorno scoprirà se appartiene o meno a $T_u$.
- **Complessità**: $O(n)$ , ogni nodo è visitato al massimo una volta
```python
def sotto_albero(P,u):
    Vis = [-1]*n
    Vis[u] = 1

    for v in P:
        if(Vis[v] == -1):
            visita(P,v,Vis)
    
    T_u = []
    for v in Vis:
        if(Vis[v] == 1): T_u.append(v)
    
    return T_u

def visita(P,v,Vis):
     
    if(Vis[P[v]] != -1):
        Vis[v] = Vis[P[v]]
        return Vis[v]

    r = 0
    if(P[v] != v):
        r = visita(P,P[v],Vis)

    Vis[v] = r
    return r
```

**Comment by CiottoloMaggico** [comment]
![immagine](https://github.com/user-attachments/assets/939ee60d-d834-4e0c-b676-28bd1877a1d2)

**Comment by daromacarlo** [comment]
![immagine](https://github.com/user-attachments/assets/55ee0d3b-3dcc-413a-b256-90376cb8400b)

**Comment by cek753** [comment]
<img width="1309" height="1430" alt="image" src="https://github.com/user-attachments/assets/b84fab8c-512b-43ca-a348-393ecd69a882" />
